In [1]:
import pandas as pd 
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
from sklearn import metrics
# read training data & test data 
df = pd.read_csv("dataset/spam.csv")

In [2]:
df.head(10)

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
5     spam  FreeMsg Hey there darling it's been 3 week's n...
6      ham  Even my brother is not like to speak with me. ...
7      ham  As per your request 'Melle Melle (Oru Minnamin...
8     spam  WINNER!! As a valued network customer you have...
9     spam  Had your mobile 11 months or more? U R entitle...

In [3]:
df.describe()

Category                 Message
count      5572                    5572
unique        2                    5157
top         ham  Sorry, I'll call later
freq       4825                      30

In [4]:
df.groupby('Category').describe()

Message                                                            \
           count unique                                                top   
Category                                                                     
ham         4825   4516                             Sorry, I'll call later   
spam         747    641  Please call our customer service representativ...   

               
         freq  
Category       
ham        30  
spam        4

In [5]:
df['label'] = df.Category.map({'ham':0, 'spam':1})

In [6]:
df.head()

Category                                            Message  label
0      ham  Go until jurong point, crazy.. Available only ...      0
1      ham                      Ok lar... Joking wif u oni...      0
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...      1
3      ham  U dun say so early hor... U c already then say...      0
4      ham  Nah I don't think he goes to usf, he lives aro...      0

In [7]:
X = df.Message
y = df.label
print(X.shape)
print(y.shape)

(5572,)
(5572,)


In [8]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(4179,)
(1393,)
(4179,)
(1393,)


In [9]:
vect = CountVectorizer()
tfidf=TfidfTransformer()

In [10]:
X_train=vect.fit_transform(X_train)
X_train_tfidf=tfidf.fit_transform(X_train)

In [11]:
X_train_tfidf.shape

(4179, 7453)

In [12]:
from sklearn.naive_bayes import MultinomialNB

clf= MultinomialNB().fit(X_train_tfidf, y_train)

In [13]:
X_test=vect.transform(X_test)
X_test_tfidf=tfidf.transform(X_test)

In [14]:
X_test_tfidf.shape

(1393, 7453)

In [15]:
predicted=clf.predict(X_test_tfidf)

In [16]:
from sklearn import metrics 

from sklearn.metrics import accuracy_score

print("Accuracy: ",accuracy_score(y_test,predicted))

print("Confusion Matrix: ",metrics.confusion_matrix(y_test,predicted))

Accuracy:  0.9641062455132807
Confusion Matrix:  [[1208    0]
 [  50  135]]


with tfidf:\
    Accuracy:  0.9641062455132807\
Confusion Matrix:  [[1208    0]\
 [  50  135]]

without tfidf: \
Accuracy:  0.9877961234745154 \
Confusion Matrix:  [[1203    5]\
 [  12  173]]